File: leaderboard/personal/go/request.go
package main

import (
	"net/http"
	"io"
    "log"
)

// Adapted from https://stackoverflow.com/a/40643195/
func OnPage(link string)(string) {
    res, err := http.Get(link)
    if err != nil {
        log.Fatal(err)
    }
    content, err := io.ReadAll(res.Body)
    res.Body.Close()
    if err != nil {
        log.Fatal(err)
    }
    return string(content)
}

func AddReqHeaders(req *http.Request, sessionCookie string)(*http.Request) {
	// Add request headers to make it look like I'm using a browser
	req.Header.Set("User-Agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:120.0) Gecko/20100101 Firefox/120.0")
	req.Header.Set("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8")
	// req.Header.Set("Accept-Encoding", "gzip, deflate, br")
	req.Header.Set("Accept-Language", "en-US,en;q=0.5")
	req.Header.Set("Connection", "keep-alive")
	req.Header.Set("DNT", "1")
	req.Header.Set("Host", "adventofcode.com")
	req.Header.Set("Sec-Fetch-Dest", "document")
	req.Header.Set("Sec-Fetch-Mode", "navigate")
	req.Header.Set("Sec-Fetch-Site", "cross-site")
	req.Header.Set("Sec-GPC", "1")
	req.Header.Set("Upgrade-Insecure-Requests", "1")

	// Add the all-important session token
	req.AddCookie(&http.Cookie{
		Name:  "session",
		Value: sessionCookie,
	})
	
	return req
}

File: leaderboard/personal/rs/src/request.rs
use super::AOC_YEAR;
use reqwest::{header::HeaderMap, Client};

pub async fn pull_personal_stats(session_cookie: String) -> String {
    // Construct URI
    let url = format!("https://adventofcode.com/{}/leaderboard/self", *AOC_YEAR);

    // Set session cookie
    let mut headers = HeaderMap::new();
    headers.insert(
        "Cookie",
        format!("session={}", session_cookie).parse().unwrap(),
    );

    // Make request
    let client = Client::new();
    let res = client
        .get(url)
        .headers(headers)
        .send()
        .await
        .expect("Failed to access URL for workplace leaderboard");

    let body = res
        .text()
        .await
        .expect("Failed to retrieve body of response");

    body
}
